pallets/capacity/src/lib.rs

```diff
@@ -475,3 +479,91 @@ impl<T: Config> Pallet<T> {
 		total_capacity.saturating_sub(rate.mul_ceil(total_capacity))
 	}
 }
+
+impl<T: Config> Nontransferable for Pallet<T> {
+	type Balance = BalanceOf<T>;
+
+	fn balance(msa_id: MessageSourceId) -> Result<Self::Balance, DispatchError> {
+		let capacity_details =
+			Self::get_capacity_for(msa_id).ok_or(Error::<T>::TargetCapacityNotFound)?;
+		Ok(capacity_details.remaining)
+	}
+
+	fn reduce_balance(msa_id: MessageSourceId, amount: Self::Balance) -> Result<(), DispatchError> {
+		let mut capacity_details =
+			Self::get_capacity_for(msa_id).ok_or(Error::<T>::TargetCapacityNotFound)?;
+
+		capacity_details.remaining = capacity_details
+			.remaining
+			.checked_sub(&amount)
+			.ok_or(Error::<T>::InsufficientBalance)?;
+
+		Self::set_capacity_for(msa_id, capacity_details);
+		Ok(())
+	}
+
+	fn increase_balance(
+		msa_id: MessageSourceId,
+		amount: Self::Balance,
+	) -> Result<(), DispatchError> {
+		let mut capacity_details =
+			Self::get_capacity_for(msa_id).ok_or(Error::<T>::TargetCapacityNotFound)?;
+
+		let new_balance = capacity_details
+			.remaining
+			.checked_add(&amount)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		ensure!(
+			new_balance <= capacity_details.total_available,
+			Error::<T>::IncreaseExceedsAvailable
+		);
+
+		capacity_details.remaining = new_balance;
+
+		Self::set_capacity_for(msa_id, capacity_details);
+
+		Ok(())
+	}
+
+	fn issue(msa_id: MessageSourceId, amount: Self::Balance) -> Result<(), DispatchError> {
+		let mut capacity_details =
+			Self::get_capacity_for(msa_id).ok_or(Error::<T>::TargetCapacityNotFound)?;
+
+		capacity_details.total_available = capacity_details
+			.total_available
+			.checked_add(&amount)
+			.ok_or(ArithmeticError::Overflow)?;
+
+		Self::set_capacity_for(msa_id, capacity_details);
+		Ok(())
+	}
+}
+
+impl<T: Config> Replenishable for Pallet<T> {
+	type Balance = BalanceOf<T>;
+
+	fn replenish_all_for_account(msa_id: MessageSourceId) -> Result<(), DispatchError> {
+		let mut capacity_details =
+			Self::get_capacity_for(msa_id).ok_or(Error::<T>::TargetCapacityNotFound)?;
+
+		capacity_details.remaining = capacity_details.total_available;
+
+		Self::set_capacity_for(msa_id, capacity_details);
+
+		Ok(())
+	}
+
+	fn replenish_by_account(
```

Comment:
Does anyone know how to write an "expect to have called" `increase_balance` test?